### ISSUE TITLE
[DocDB] flushed_op_id for IntentsDB could become behind log_min_seconds_to_retain time window and cause a gap during local bootstrap

### ISSUE THREAD

**Issue by ttyusupov** [yugabyte/yugabyte-db#16684]
Jira Link: [DB-6057](https://yugabyte.atlassian.net/browse/DB-6057)
### Description

Flushed_op_id for IntentsDB could become behind `log_min_seconds_to_retain` time window and this can cause the following:

1) Local bootstrap will have to replay all WAL files RBS has downloaded (and it can download way more than just for 15 minutes, because WAL files could be generated during RBS) and therefore it will take long time. Also it can cause RBS session expiration on the source node (#15877) that might lead to the situation with stuck Raft config. 
2) Log message from local bootstrap that first WAL segment is later than flushed_op_id for intents DB: `Segment <...> cannot be used as the first segment to start replay with according to our OpId and retention criteria. However, this is already the earliest segment so we have to start replay here. We should probably investigate how we got into this situation.`. Some (but not all) cases from #6112 are exactly because of this.

### Warning: Please confirm that this issue does not contain any sensitive information

- [X] I confirm this issue does not contain any sensitive information.

[DB-6057]: https://yugabyte.atlassian.net/browse/DB-6057?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by mbautin** [comment]
Here is an idea that @ttyusupov and I discussed:

Modify the WAL footer to store the op id of the latest committed operation for each operation type: intents RocksDB writes, regular RocksDB writes, and metadata change operations that affect the tablet superblock. This information will be used to determine whether the operations of a specific type need to be replayed from a WAL segment during the bootstrap process.

Update the WAL writing process to include the op id of the latest committed operation of each type, up to and including the end of the current WAL segment. This ensures that the footer contains the most recent information about committed operations for each operation type.

Modify the bootstrap process to use the stored op ids in the WAL footer to determine which operations need to be replayed. During bootstrapping, read the last WAL footer and compare the last committed op id for each operation type with the last flushed op id (as indicated by the flushed frontier of RocksDB metadata for intents and regular RocksDB writes, and by the op id of the latest change metadata operation recorded in the tablet superblock). If they are equal, there is no need to replay operations of that type from that segment. Otherwise, those operations need to be replayed.

**Comment by arpang** [comment]
@mbautin, if an intent is written, followed by a long period of non-transactional writes, is there an existing mechanism by which the intentsdb memtable will get flushed even though it is not full? If not, do you think we will have to introduce it? Because without it, on a restart, we would still need to replay many segments just to reapply that one intent write.

**Comment by mbautin** [comment]
We can make the determination of whether to replay a WAL segment for each segment separately based on that segment's metadata of last committed op id of each operation type. The new intent will only affect the last Raft-committed op id of the intent type recorded in the WAL segment that the intent written operation is written to, as well as all subsequent WAL segments. Earlier segments would not have to be replayed because we know that they do not contain new un-flushed intent write operations. Similar logic applies to metadata change operations (superblock updates).

**Comment by mbautin** [comment]
Sorry, my answer was for the opposite situation, of a long period of non-transactional writes, followed by a single intent.

**Comment by mbautin** [comment]
@arpang yes, it would be good to have a mechanism to flush the intents once in a while even if the memtable is not full. We still need to maintain the ordering guarantee that the regular RocksDB memtable containing effects of the same Raft operation is flushed first, as we do today, to avoid losing intents applied to regular RocksDB and deleted from intents RocksDB.

**Comment by mbautin** [comment]
@ttyusupov has also mentioned this optimization: even if we don't flush the intents and we have an old WAL segment with an intent followed by many WAL segments with non-transactional writes, we should be able to mostly skip replaying these newer WAL segments. After replaying the intent write operation from the oldest segment, we will update the flushed op id of the last intent in memory, and it will be equal to the last committed intent op id stored in footers of segments that follow, allowing us to skip replaying operations from those segments (assuming that all the regular RocksDB operations in those segments are flushed too).

**Comment by arpang** [comment]
@mbautin Makes sense for replay. But we can still end up retaining a large number of WAL segments. Flushing intents once in a while, even if the memtable is not full should help limit retention.

Alternatively, similar to how @ttyusupov suggests replaying only the necessary segments selectively, we can retain segments selectively creating holes in between (for instance seg0 and seg2 are retained, seg1 is not), but perhaps it is unnecessary complexity (?).

**Comment by ttyusupov** [comment]
I was able to reproduce this for YCQL (and we met this on real cluster with YCQL) with a test on local cluster:
```
export TS_FLAGS='"memstore_size_mb=1","log_min_seconds_to_retain=20","log_max_seconds_to_retain=20","retryable_request_timeout_secs=15","log_segment_size_bytes=102400","initial_log_segment_size_bytes=102400"'

./bin/yb-ctl --rf=1 start --tserver_flags $TS_FLAGS

./bin/ycqlsh -e "CREATE KEYSPACE ks"
./bin/ycqlsh -e "CREATE TABLE ks.test (k int PRIMARY KEY, v int) WITH tablets = 1 AND transactions = { 'enabled' : true };"

# Should trigger intents DB flush
python3 -c "for i in range(1, 10000): print('BEGIN TRANSACTION INSERT INTO ks.test (k, v) VALUES(', i, ', 1); END TRANSACTION;')" | ./bin/ycqlsh

# Should trigger only regular DB flush and latest flushed write to intents DB will be earlier than log_max_seconds_to_retain ago
sleep 20 && python3 -c "for i in range(10000, 30000): print(i, ',', 2)" | ./bin/ycqlsh -e "COPY ks.test (k, v) FROM STDIN"

# After first restart we will have empty memtables and therefore Raft log GC will not take flushed_op_ids into account
./bin/yb-ctl --rf=1 restart --tserver_flags $TS_FLAGS

# Restart again to reproduce gap between flushed_op_ids and WAL during local bootstrap
./bin/yb-ctl --rf=1 restart --tserver_flags $TS_FLAGS

# Check local bootstrap logs
```

Result:
```
I0201 19:18:42.771219 1418404 tablet_bootstrap.cc:388] T 41fd0bc15f0c4217ae362b4f4c0a27ce P b4b5befd043e49e9a1d989de8fedc3a1: Bootstrap optimizer (skip_flushed_entries): op_id_replay_lowest=1.19998 (regular_op_id=1.20999, intents_op_id=1.19998, has_intents_db=1)
...
I0201 19:18:42.774618 1418404 tablet_bootstrap.cc:1243] T 41fd0bc15f0c4217ae362b4f4c0a27ce P b4b5befd043e49e9a1d989de8fedc3a1: Segment /nfusr/dev-server/timur/yugabyte-data/node-1/disk-1/yb-data/tserver/wals/table-cc6a38719daa4b7abba6644114ecea55/tablet-41fd0bc15f0c4217ae362b4f4c0a27ce/wal-000000045 cannot be used as the first segment to start replay with according to our OpId and retention criteria. However, this is already the earliest segment so we have to start replay here. We should probably investigate how we got into this situation. op_id=1.20400, op_id_replay_lowest=1.19998, first_op_time=8318742731.886s, min_seconds_to_retain_logs=15.000s, replay_from_this_or_earlier_time_was_initialized=1, *replay_from_this_or_earlier_time=8318742730.980s, is_first_op_id_low_enough=0, is_first_op_time_early_enough=0
```

**Comment by ttyusupov** [comment]
Although there is a commit https://github.com/yugabyte/yugabyte-db/commit/3ea47159cc55e665352d46e6f4fa478b7aff1ffd, the scenario above still can happen because even when RocksDB flush is forced frontier is not updated when memtable is empty, since there is nothing to flush.

**Comment by ttyusupov** [comment]
Also need to check whether table truncate also lead to this gap, since we wipe both rocksdbs and there will be no frontiers.
